### PR TITLE
Run relevant CI tests based on what's changed in the ChangeList

### PR DESCRIPTION
**Pull request by anandhkb**
cpp_changed==cpp+python+notebooks, python_changed==python+notebooks, notebooks_changed==notebooks

### CODE REVIEW

File: ci/test.sh
Comment (rlratzel) [inline ci/test.sh:53]: I think this code would be more appropriate in `ci/gpu/build.sh` and have the variables exported and then referenced in the lower-level scripts (like you're doing for the `test-notebooks.sh` script).  The main reason being that `test.sh` is sometimes also used by devs outside of CI, and this code would prevent that due to how it's trying to access info about a PR.  It would also mean you wouldn't have to `source test.sh` (which can sometimes have unintended side-effects) since the variables would be set in the top-level calling script.

Once this code is moved out, `test.sh` should be written to also handle cases where the various `python_changed`, `nb_changed`, etc. variables are not set, so devs can continue to run `test.sh` outside of CI.

File: ci/test.sh
Comment (rlratzel) [inline ci/test.sh:38]: It might be better to make the variable names a little more accurate, since `python_changed` implies python code changed when it could actually be set by way of a C++ code change.  Perhaps the naming convention could be `run_python_tests`, `run_nb_tests`, etc.?

File: ci/gpu/build.sh
Comment (rlratzel) [inline ci/gpu/build.sh:160]: I believe the codecov step below will fail if python tests aren't run, since the python tests generate the output needed by codecov.  You might try adding something to this `if` block to prevent codecov from running if python tests did not run.

File: ci/test.sh
Comment (rlratzel) [inline ci/test.sh:54]: Based on the CI failures, it looks like this isn't being set properly beause `CUGRAPH_ROOT` isn't being set properly, because (I think) this script is now being `source`'d .  I think my other suggestion would eliminate the need to `source` this script and should fix that problem.

**Review by rlratzel — CHANGES_REQUESTED** [review]
Thanks for doing this.  I have a few suggestions and change requests to hopefully fix the CI failures.

File: ci/test.sh
Comment (rlratzel) [inline ci/test.sh:53]: Somewhere around here might be a good place to print a summary of what's going to be run, what's going to be skipped, and (most importantly) why, since it might be confusing to devs who are expecting certain tests to run which end up getting skipped.

**Review by rlratzel — CHANGES_REQUESTED** [review]

File: ci/test.sh
Comment (anandhkb) [reply ci/test.sh:53]: Got it, thanks Rick

File: ci/test.sh
Comment (anandhkb) [reply ci/test.sh:38]: Make sense

File: ci/gpu/build.sh
Comment (anandhkb) [reply ci/gpu/build.sh:160]: Got it, thanks Rick

File: ci/test.sh
Comment (anandhkb) [reply ci/test.sh:54]: Sounds good

File: ci/test.sh
Comment (anandhkb) [reply ci/test.sh:53]: Fair ask, will do

File: ci/gpu/build.sh
Comment (anandhkb) [reply ci/gpu/build.sh:124]: Thanks @ajschmidt8 , how to distinguish the nightly builds from the PR builds then? Based on that, I can add an 'if condition' and hide the new code under that 'if block'.

Comment (anandhkb) [reply ci/gpu/build.sh:124]: if [ "$BUILD_MODE" = "pull-request" ]; then
<new code>
fi

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]
Left some suggestions. Also, I don't see the `doc_changed` value being used anywhere throughout the PR. Can this be removed?

File: ci/gpu/build.sh
Comment (anandhkb) [reply ci/gpu/build.sh:131]: @ajschmidt8 @rlratzel 
RE: "First, the else statement can be removed as long as the proper defaults values are set above the if statement"

That's wrong logic. The logic we want to implement is:
a. If the build_mode is NOT a pull-request, then run ALL tests (cpp_tests = true, python_tests=true, nb_tests=true)
b. Else (build_mode is a pull-request), then initialize cpp_tests = false, python_tests=false, nb_tests=false
c. Walk through the filenames changed by the ChangeList and based on what is changed, update the cpp_tests/python_tests/nb_tests variables to 'true'

Comment (ajschmidt8) [reply ci/gpu/build.sh]: @anandhkb, I understand what you're saying. the two example code blocks below are the same however. We can eliminate the `else` statement by setting the default values for the `run_*_tests` variables above the `if` block.

```sh
if [ "$BUILD_MODE" == "pull-request" ]; then
    run_cpp_tests="false" run_python_tests="false" run_nb_tests="false"
else
    run_cpp_tests="true" run_python_tests="true" run_nb_tests="true"
fi
```

```sh
run_cpp_tests="true" run_python_tests="true" run_nb_tests="true"
if [ "$BUILD_MODE" == "pull-request" ]; then
    run_cpp_tests="false" run_python_tests="false" run_nb_tests="false"
fi
```

Comment (anandhkb) [reply ci/gpu/build.sh]: Yes, you are absolutely right @ajschmidt8 and my latest update addresses your point.

**Review by rlratzel — APPROVED** [review]
Looks good. I have one suggestion but it need not hold up my approval.

**Review by ajschmidt8 — APPROVED** [review]
Approved pending the resolution of Rick's comment

File: ci/gpu/build.sh
Comment (rlratzel) [reply ci/gpu/build.sh:145]: I would personally err on the side of running the tests rather than skipping them when in doubt. If you enumerate every subdir and file that triggers testing, and we add a subdir or file that isn't listed, the tests won't get run.  By explicitly listing the cpp dir then only items to exclude, you don't risk missing test runs on newly added source dirs or files (for instance, in the past few months we added libcugraph_etl). Missing test runs could potentially go unnoticed for a while in that situation and lead to a broken upstream.

Comment (anandhkb) [reply ci/gpu/build.sh]: You are correct Rick and I have update the commit based on your advise

**Comment by anandhkb** [comment]
@gpucibot merge

**Pull request merged**
